src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -83,6 +86,189 @@ object TaskGroupBuilder extends StrictLogging {
+    val disk = BigDecimal(pod.executorResources.disk) + pod.containers.map(c => BigDecimal(c.resources.disk)).sum
+    val gpus = BigDecimal(pod.executorResources.gpus) + pod.containers.map(c => BigDecimal(c.resources.gpus)).sum
+
+    val matchedCpus = matchedResources.filter(_.getName == "cpus").map(c => BigDecimal(c.getScalar.getValue)).sum
+    val matchedMem = matchedResources.filter(_.getName == "mem").map(c => BigDecimal(c.getScalar.getValue)).sum
+    val matchedDisk = matchedResources.filter(_.getName == "disk").map(c => BigDecimal(c.getScalar.getValue)).sum
+    val matchedGpus = matchedResources.filter(_.getName == "gpus").map(c => BigDecimal(c.getScalar.getValue)).sum
+
+    assert(cpus <= matchedCpus)
+    assert(mem <= matchedMem)
+    assert(disk <= matchedDisk)
+    assert(gpus <= matchedGpus)
+  }
+
+  // We have list of (BigDecimal, Resource) pairs in order to avoid using == and != operators with doubles.
+  private case class Resources(
+      cpus: List[(BigDecimal, mesos.Resource)],
+      mem: List[(BigDecimal, mesos.Resource)],
+      disk: List[(BigDecimal, mesos.Resource)],
+      gpus: List[(BigDecimal, mesos.Resource)])
+
+  private class ResourceConsumer(initial: Resources) {
+    private var cpus = initial.cpus
+    private var mem = initial.mem
+    private var disk = initial.disk
+    private var gpus = initial.gpus
+
+    // Find a CPU resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeCpus(quantity: Double): Option[mesos.Resource] = {
+      consume(cpus, BigDecimal(quantity)).map {
+        case (resources, resource) =>
+          cpus = resources
+          resource
+      }
+    }
+
+    // Find a MEM resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeMem(quantity: Double): Option[mesos.Resource] = {
+      consume(mem, BigDecimal(quantity)).map {
+        case (resources, resource) =>
+          mem = resources
+          resource
+      }
+    }
+
+    // Find a DISK resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeDisk(quantity: Double): Option[mesos.Resource] = {
+      consume(disk, BigDecimal(quantity)) map {
+        case (resources, resource) =>
+          disk = resources
+          resource
+      }
+    }
+
+    // Find a GPU resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeGpus(quantity: Double): Option[mesos.Resource] = {
+      consume(gpus, BigDecimal(quantity)).map {
+        case (resources, resource) =>
+          gpus = resources
+          resource
+      }
+    }
+
+    private def consume(
+      resources: List[(BigDecimal, mesos.Resource)],
+      quantity: BigDecimal): Option[(List[(BigDecimal, mesos.Resource)], mesos.Resource)] =
+      if (quantity <= BigDecimal(0)) {
+        None
+      } else {
+        resources.partition(_._1 != quantity) match {
```

Comment:
What is `_._1`? How about making this explicit?

Reply:
Done.